File: src/main/java/gg/grounds/grpc/StatusGrpcService.java
package gg.grounds.grpc;

import com.google.protobuf.Empty;
import gg.grounds.domain.model.MotdEntity;
import gg.grounds.domain.service.HeartbeatService;
import gg.grounds.domain.service.MaintenanceService;
import gg.grounds.domain.service.MotdService;
import gg.grounds.grpc.status.*;
import gg.grounds.infrastructure.event.StatusEventBus;
import gg.grounds.mapper.ProtoMapper;
import gg.grounds.application.*;
import io.quarkus.grpc.GrpcService;
import io.smallrye.mutiny.Multi;
import io.smallrye.mutiny.Uni;
import jakarta.inject.Inject;

import java.time.Instant;
import java.util.UUID;

@GrpcService
public class StatusGrpcService extends MutinyStatusServiceGrpc.StatusServiceImplBase {

    @Inject
    MotdService motdService;
    @Inject
    MaintenanceService maintenanceService;
    @Inject
    HeartbeatService heartbeatService;
    @Inject
    SnapshotService snapshotService;

    @Inject
    StatusEventBus eventBus;
    @Inject
    ProtoMapper mapper;
    @Inject
    StatusChangePublisher publisher;


    // ----------------- STREAM -----------------

    @Override
    public Multi<StatusEvent> subscribeStatus(SubscribeStatusRequest request) {
        Multi<StatusEvent> initial = Multi.createFrom().uni(
                snapshotService.buildEvent(StatusEventType.STATUS_EVENT_TYPE_SNAPSHOT)
        );

        return Multi.createBy().concatenating().streams(
                initial,
                eventBus.stream()
        );
    }

    // ----------------- MOTD -----------------

    @Override
    public Uni<Motd> createMotd(CreateMotdRequest request) {
        MotdEntity e = new MotdEntity();
        e.name = request.getName();
        e.payloadJson = request.getPayloadJson();
        e.priority = request.getPriority();
        e.startsAt = mapper.fromTs(request.getStartsAt());
        e.endsAt = mapper.fromTs(request.getEndsAt());
        e.enabled = request.getEnabled();

    return motdService.create(e)
            .map(mapper::toProto)
            .invoke(publisher::motdChanged);
  }

    @Override
    public Uni<Motd> updateMotd(UpdateMotdRequest request) {
        UUID id = UUID.fromString(request.getId());

        MotdEntity e = new MotdEntity();
        e.name = request.getName();
        e.payloadJson = request.getPayloadJson();
        e.priority = request.getPriority();
        e.startsAt = mapper.fromTs(request.getStartsAt());
        e.endsAt = mapper.fromTs(request.getEndsAt());
        e.enabled = request.getEnabled();

        return motdService.update(id, e)
                .map(mapper::toProto)
                .invoke(publisher::motdChanged);
    }

    @Override
    public Uni<Motd> getMotd(GetMotdRequest request) {
        return motdService.get(UUID.fromString(request.getId()))
                .map(mapper::toProto);
    }

    @Override
    public Uni<Empty> deleteMotd(DeleteMotdRequest request) {
        return motdService.delete(UUID.fromString(request.getId()))
                .replaceWith(Empty.getDefaultInstance())
                .invoke(publisher::motdChanged);
    }

    @Override
    public Uni<ListMotdsResponse> listMotds(ListMotdsRequest request) {
        return motdService.list(request.getIncludeDisabled())
                .map(list -> {
                    ListMotdsResponse.Builder b = ListMotdsResponse.newBuilder();
                    for (var e : list) b.addMotds(mapper.toProto(e));
                    return b.build();
                });
    }

    @Override
    public Uni<GetActiveMotdResponse> getActiveMotd(GetActiveMotdRequest request) {
        Instant now = mapper.fromTs(request.getNow());
        return motdService.getActive(now)
                .map(active -> GetActiveMotdResponse.newBuilder()
                        .setMotd(mapper.toProto(active))
                        .build());
    }

    // ----------------- MAINTENANCE -----------------


    @Override
    public Uni<GetMaintenanceResponse> getMaintenance(GetMaintenanceRequest request) {
        return maintenanceService.getConfig()
                .map(cfg -> {
                    boolean effective = maintenanceService.isEffectiveEnabled(cfg, Instant.now());
                    return GetMaintenanceResponse.newBuilder()
                            .setConfig(mapper.toProto(cfg, effective))
                            .build();
                });
    }

    @Override
    public Uni<GetMaintenanceResponse> setMaintenance(SetMaintenanceRequest request) {
        Instant starts = mapper.fromTs(request.getStartsAt());
        Instant ends = mapper.fromTs(request.getEndsAt());

        return maintenanceService.setConfig(request.getEnabled(), request.getMessage(), starts, ends)
                .map(cfg -> {
                    boolean effective = maintenanceService.isEffectiveEnabled(cfg, Instant.now());
                    return GetMaintenanceResponse.newBuilder()
                            .setConfig(mapper.toProto(cfg, effective))
                            .build();
                })
                .invoke(publisher::maintenanceChanged);
    }

    @Override
    public Uni<WhitelistEntry> addWhitelist(AddWhitelistRequest request) {
        UUID uuid = UUID.fromString(request.getUuid());
        return maintenanceService.addWhitelist(uuid, request.getNote())
                .map(mapper::toProto);
    }

    @Override
    public Uni<Empty> removeWhitelist(RemoveWhitelistRequest request) {
        UUID uuid = UUID.fromString(request.getUuid());
        return maintenanceService.removeWhitelist(uuid)
                .replaceWith(Empty.getDefaultInstance());
    }

    @Override
    public Uni<ListWhitelistResponse> listWhitelist(ListWhitelistRequest request) {
        return maintenanceService.listWhitelist()
                .map(list -> {
                    ListWhitelistResponse.Builder b = ListWhitelistResponse.newBuilder();
                    for (var e : list) b.addEntries(mapper.toProto(e));
                    return b.build();
                });
    }

    // ----------------- HEARTBEAT -----------------


    @Override
    public Uni<SendHeartbeatResponse> sendHeartbeat(SendHeartbeatRequest request) {
        ProxyHeartbeat hb = request.getHeartbeat();
        String metaJson = "{}";

        if (hb.getMetaCount() > 0) {
            StringBuilder sb = new StringBuilder("{");
            boolean first = true;
            for (var entry : hb.getMetaMap().entrySet()) {
                if (!first) sb.append(",");
                first = false;
                sb.append("\"").append(escapeJson(entry.getKey())).append("\":")
                        .append("\"").append(escapeJson(entry.getValue())).append("\"");
            }
            sb.append("}");
            metaJson = sb.toString();
        }

        return heartbeatService.upsert(
                hb.getProxyId(),
                hb.getPlayerCount(),
                hb.getHealthy(),
                hb.getVersion(),
                metaJson
        )
                .map(ignored -> SendHeartbeatResponse.newBuilder()
                        .setReceivedAt(mapper.ts(Instant.now()))
                        .build());
    }

    private static String escapeJson(String s) {
        if (s == null) return "";
        return s.replace("\\", "\\\\").replace("\"", "\\\"");
    }
}

File: src/main/kotlin/gg/grounds/grpc/StatusGrpcService.kt
package gg.grounds.grpc

import com.google.protobuf.Empty
import gg.grounds.application.SnapshotService
import gg.grounds.application.StatusChangePublisher
import gg.grounds.domain.model.MaintenanceConfigEntity
import gg.grounds.domain.model.MaintenanceWhitelistEntity
import gg.grounds.domain.model.MotdEntity
import gg.grounds.domain.model.ProxyHeartbeatEntity
import gg.grounds.domain.service.HeartbeatService
import gg.grounds.domain.service.MaintenanceService
import gg.grounds.domain.service.MotdService
import gg.grounds.grpc.status.*
import gg.grounds.infrastructure.event.StatusEventBus
import gg.grounds.mapper.ProtoMapper
import io.quarkus.grpc.GrpcService
import io.smallrye.mutiny.Multi
import io.smallrye.mutiny.Uni
import jakarta.inject.Inject
import java.time.Instant
import java.util.*
import java.util.function.Function

@GrpcService
class StatusGrpcService : MutinyStatusServiceGrpc.StatusServiceImplBase() {
    @Inject
    lateinit var motdService: MotdService

    @Inject
    lateinit var maintenanceService: MaintenanceService

    @Inject
    lateinit var heartbeatService: HeartbeatService

    @Inject
    lateinit var snapshotService: SnapshotService

    @Inject
    lateinit var eventBus: StatusEventBus

    @Inject
    lateinit var mapper: ProtoMapper

    @Inject
    lateinit var publisher: StatusChangePublisher


    // ----------------- STREAM -----------------
    override fun subscribeStatus(request: SubscribeStatusRequest?): Multi<StatusEvent?>? {
        val initial = Multi.createFrom().uni<StatusEvent?>(
            snapshotService.buildEvent(StatusEventType.STATUS_EVENT_TYPE_SNAPSHOT)
        )

        return Multi.createBy().concatenating().streams<StatusEvent?>(
            initial,
            eventBus.stream()
        )
    }

    // ----------------- MOTD -----------------
    override fun createMotd(request: CreateMotdRequest): Uni<Motd?>? {
        val e = MotdEntity()
        e.name = request.getName()
        e.payloadJson = request.getPayloadJson()
        e.priority = request.getPriority()
        e.startsAt = mapper.fromTs(request.getStartsAt())
        e.endsAt = mapper.fromTs(request.getEndsAt())
        e.enabled = request.getEnabled()

        return motdService.create(e)!!
            .map<Motd?>(Function { e: MotdEntity? -> mapper.toProto(e) })
            .invoke(Runnable { publisher.motdChanged() })
    }

    override fun updateMotd(request: UpdateMotdRequest): Uni<Motd?>? {
        val id = UUID.fromString(request.getId())

        val e = MotdEntity()
        e.name = request.getName()
        e.payloadJson = request.getPayloadJson()
        e.priority = request.getPriority()
        e.startsAt = mapper.fromTs(request.getStartsAt())
        e.endsAt = mapper.fromTs(request.getEndsAt())
        e.enabled = request.getEnabled()

        return motdService.update(id, e)!!
            .map<Motd?>(Function { e: MotdEntity? -> mapper.toProto(e) })
            .invoke(Runnable { publisher.motdChanged() })
    }

    override fun getMotd(request: GetMotdRequest): Uni<Motd?>? {
        return motdService.get(UUID.fromString(request.getId()))!!
            .map<Motd?>(Function { e: MotdEntity? -> mapper.toProto(e) })
    }

    override fun deleteMotd(request: DeleteMotdRequest): Uni<Empty?>? {
        return motdService.delete(UUID.fromString(request.getId()))!!
            .replaceWith<Empty?>(Empty.getDefaultInstance())
            .invoke(Runnable { publisher.motdChanged() })
    }

    override fun listMotds(request: ListMotdsRequest): Uni<ListMotdsResponse?>? {
        return motdService.list(request.getIncludeDisabled())!!
            .map<ListMotdsResponse?>(Function { list: MutableList<MotdEntity?>? ->
                val b = ListMotdsResponse.newBuilder()
                for (e in list!!) b.addMotds(mapper.toProto(e))
                b.build()
            })
    }

    override fun getActiveMotd(request: GetActiveMotdRequest): Uni<GetActiveMotdResponse?>? {
        val now = mapper.fromTs(request.getNow())
        return motdService.getActive(now)!!
            .map<GetActiveMotdResponse?>(Function { active: MotdEntity? ->
                GetActiveMotdResponse.newBuilder()
                    .setMotd(mapper.toProto(active))
                    .build()
            })
    }


    // ----------------- MAINTENANCE -----------------
    override fun getMaintenance(request: GetMaintenanceRequest?): Uni<GetMaintenanceResponse?>? {
        return maintenanceService.config!!
            .map<GetMaintenanceResponse?>(Function { cfg: MaintenanceConfigEntity? ->
                val effective = maintenanceService.isEffectiveEnabled(cfg, Instant.now())
                GetMaintenanceResponse.newBuilder()
                    .setConfig(mapper.toProto(cfg, effective))
                    .build()
            })
    }

    override fun setMaintenance(request: SetMaintenanceRequest): Uni<GetMaintenanceResponse?>? {
        val starts = mapper.fromTs(request.getStartsAt())
        val ends = mapper.fromTs(request.getEndsAt())

        return maintenanceService.setConfig(request.getEnabled(), request.getMessage(), starts, ends)!!
            .map<GetMaintenanceResponse?>(Function { cfg: MaintenanceConfigEntity? ->
                val effective = maintenanceService.isEffectiveEnabled(cfg, Instant.now())
                GetMaintenanceResponse.newBuilder()
                    .setConfig(mapper.toProto(cfg, effective))
                    .build()
            })
            .invoke(Runnable { publisher.maintenanceChanged() })
    }

    override fun addWhitelist(request: AddWhitelistRequest): Uni<WhitelistEntry?>? {
        val uuid = UUID.fromString(request.getUuid())
        return maintenanceService.addWhitelist(uuid, request.getNote())
            .map<WhitelistEntry?>(Function { e: MaintenanceWhitelistEntity -> mapper.toProto(e) })
    }

    override fun removeWhitelist(request: RemoveWhitelistRequest): Uni<Empty?>? {
        val uuid = UUID.fromString(request.getUuid())
        return maintenanceService.removeWhitelist(uuid)!!
            .replaceWith<Empty?>(Empty.getDefaultInstance())
    }

    override fun listWhitelist(request: ListWhitelistRequest?): Uni<ListWhitelistResponse?>? {
        return maintenanceService.listWhitelist()!!
            .map<ListWhitelistResponse?>(Function { list: MutableList<MaintenanceWhitelistEntity?>? ->
                val b = ListWhitelistResponse.newBuilder()
                for (e in list!!) b.addEntries(mapper.toProto(e))
                b.build()
            })
    }


    // ----------------- HEARTBEAT -----------------
    override fun sendHeartbeat(request: SendHeartbeatRequest): Uni<SendHeartbeatResponse?>? {
        val hb = request.getHeartbeat()
        var metaJson = "{}"

        if (hb.getMetaCount() > 0) {
            val sb = StringBuilder("{")
            var first = true
            for (entry in hb.getMetaMap().entries) {
                if (!first) sb.append(",")
                first = false
                sb.append("\"").append(escapeJson(entry.key)).append("\":")
                    .append("\"").append(escapeJson(entry.value)).append("\"")
            }
            sb.append("}")
            metaJson = sb.toString()
        }

        return heartbeatService.upsert(
            hb.getProxyId(),
            hb.getPlayerCount(),
            hb.getHealthy(),
            hb.getVersion(),
            metaJson
        )!!
            .map<SendHeartbeatResponse?>(Function { ignored: ProxyHeartbeatEntity? ->
                SendHeartbeatResponse.newBuilder()
                    .setReceivedAt(mapper.ts(Instant.now()))
                    .build()
            })
    }

    companion object {
        private fun escapeJson(s: String?): String {
            if (s == null) return ""
            return s.replace("\\", "\\\\").replace("\"", "\\\"")
        }
    }
}
